### ISSUE TITLE
ENH: player contracts

### ISSUE THREAD

**Issue by jpcsupplies** [jpcsupplies/Economy_mod#36]
Works in concert with the ability to buy/sell commodities..
create "delivery contracts" for freighter style players to deliver x number of goods to player y
pays on completion of delivery

waaaay future milestone enh tho

**Comment by midspace** [comment]
Hopefully you've seen the latest update notes.
http://forum.keenswh.com/threads/7368279
"autopilot improvements"
"Auto-piloted ships can now evade asteroids and ships without crashing into them."

An excerpt from part of the PB code to the Auto Pilot.
[code]
var remote = list[0] as IMyRemoteControl;
remote.ClearWaypoints();
Vector3D player;
bool success = remote.GetNearestPlayer(out player);
if (success)
{
    player = remote.GetFreeDestination(player, 200, 15.0f);
    remote.AddWaypoint(player, "Player");
    remote.SetAutoPilotEnabled(true);
}
[/code]

`GetNearestPlayer` from what I understand only works with remote blocks that are controlled by NPC.

**Comment by jpcsupplies** [comment]
Now that is handy..  now our delivery ships dont end up like a fly on a windscreen hopefully..  and any movies I make wont result in cluster F***s at a single waypoint lol

Also reminds me to update.. woops

**Comment by jpcsupplies** [comment]
This enhancement could also work as as separate workshop mod capable of being an API plugin.. sort of a proof of concept,  if it used the same connection ID..   it could access our bank file to pay players for completed missions, or bill players for posting missions..?
Combined with the improved autopilot ai and tracking; or even pirate encounters this gives almost unlimited potential for things like delivery escort,raid or bounty contracts

Do you think space marine might be interested in playing with the idea of making a mission/contract mod which can be set to talk to ours?

**Comment by Spcemarine** [comment]
No, I don't think I'm interested in making a mission/contract mod.

Just browsed through the code a bit and I must say a lot of the code seems familiar to me. Weird. I think I wrote some of it but I cannot remember pushing it to this repo... Just kidding, just treat it kindly. ;P

**Comment by jpcsupplies** [comment]
Midspace has been helping a lot so they are probably not reinventing the wheel.

Feel free to offer advice on it tho, goodness knows i need the help, i was trained in software development under ASIC, BASIC, QBASIC,  PASCAL, and Visual Basic 5.
The only C i played with was some map definitions under Lima Bean for a MUD server, and coincidental similarities in perl..   C# is properly doing my head in!  :)

**Comment by Spcemarine** [comment]
No need for reinventing the wheel. Midspace told me that the code is being used here anyway and I have no objections.

I also started programming C# with Space Engineers but I started with Java and that's very similar to C# so it wasn't very difficult and I was never trained in anything concerning software engineering so I was kind of used to teaching myself how to program in C#. So I don't really have advice and since Midspace is here there is no need for another adviser as he might be one of the best you can get in the SE community. ;)

BTT:
I wonder a to which extend mods can talk to one another. Is it possible to create some base classes in one mod that can be used in another one? Then you could just create an API for missions / contracts and ppl who want to create some for their servers might just create them based on this mod.

**Comment by midspace** [comment]
Cannot uses classes from one mod in another, as we cannot control the compile of the project and dependancies. The .Net compiler that Keen uses doesn't use any user generated project file.
That basically leaves us sharing the Message classes and the id number for communicating.

**Comment by jpcsupplies** [comment]
In so much as a mod api is concerned, mod makers making mods to talk to our mod, really only need something that allows external mods to add or remove money from a player.. which should be possible under message classes? 

 Everything else the 3rd party mod can do directly itself, no need to talk to ours...  unless someone makes a trade bot for the stock market of course.  But most economy systems I've hit in other games are basically only interacting with the player balance - take minecraft/bukkit essentials/essentials economy. It did various interesting things, but mods that talked to it, really only needed to pay (or tax) players balance.

Another method i've mentioned before is the "one hit read" method.   Have a data file (eg bank balance) that is not held open, and has a read performed on it prior to making any changes in game.
This leaves the file unlocked most of the time and can be written to externally - (eg by a website CGI script)

So any changes made outside of the game, immediately have an in-game effect.
The only difficulty of this method is:
a: your code needs some sort of on-error-file-locked, retry  logic 
b: if the file gets too big, reading it every time someone types /bal /pay /sell whatever could impact mod performance.
c: if either the mod or the external changes are too frequent, the on-error-retry-logic can tie up the mod too.
d: if the external program messes up the file; mod or external program needs to deal with that too.. which usually means a backup file.. which makes things slower again processing two files.

Advantages:
No need for any crazy changes to the Mod API system - since only the saved data is touched, not the game itself.


Midspace could use one-hit logic on the admin mod MOTD file for example, allowing admins to make changes externally that immediately show up in game.

**Comment by midspace** [comment]
Writing to the accounts file also requires know the current structure.
ergo => api.

A lot of the .net classes to make it safe to do this, just aren't whitelisted by Keen to use in the Mod Api.

It might make a little more sense to have a separate file created, that Econ reads as Input, and once read, deleted, making it transactional.
But then, how do you respond back to external requests?

But I'd rather deal with in game API, as it's easier to manage.
We just need to publish the required classes needed for communicating with our mod.
They can be wrappered for easy use. Just plug in your detail, and Invoke.

**Comment by jpcsupplies** [comment]
Still think the original post idea of contracts is worth looking into more..  at the very least a bounty system, or a delivery contract.  Or some sort of missions framework. Delivery contracts have the advantage that our range check logic can be used,  bounties have the disadvantage that we still need a way to work out HOW and WHO killed the target of the bounty.  Simply using proximity checks might open the possibility of people exploiting the bounty system...

But the seek and destroy code in autopilot does open up the reverse possibility... instead of a bounty, create a kill contract...    which launches a guided missile at the target..  rather un-blue mining sort of thing tho.

**Comment by jpcsupplies** [comment]
Would this in an in-game script give us player seeking missiles BTW ?   Probably scanrange is limited to chunk size tho.. 

An excerpt from part of the PB code to the Auto Pilot.
```
var remote = list[0] as IMyRemoteControl;
remote.ClearWaypoints();
Vector3D player;
bool success = remote.GetNearestPlayer(out player);
if (success)
{
    player = remote.GetFreeDestination(player, 200, 15.0f);
    remote.AddWaypoint(player, "Player");
    remote.SetAutoPilotEnabled(true);
}
```

**Comment by midspace** [comment]
Ship requires remote cube.
Remote cube ownership has to be set to an NPC identity.

NPC doesn't have to enemy or in different faction.

**Comment by jpcsupplies** [comment]
Ok, for the moment forget everything  wrote above.

I was shuttlng ice the other day between europa and blue mining to refill the market as it had been used up, and it was quite fun.     I was thinking about what you said in regard to my trade network hud thing using what really should be for missions..   and thought...  why dont we use it for missions then?

Something like new players (or anyone online)  get issued a "urgent order from %trade network name % for 1000 XX item."        the item it chooses for the order would be whatever item the market has the least (or none) of (usually an ore since new players can mine those)   and pay some sort of bonus if a player sells that to the market next.

From a logic perspective, this is quite easy to add to our existing /sell command as an extra check to see if a sale matches an "urgent order request"

Since urgent orders are drawn from whatever item we have least of (or none of)   these missions probably dont even need to be saved in the database, and can simply exist in the current DS session memory making things even more simple.  (after a reboot it would simply generate a new urgent order based on stock levels)   

Extended logic - remote npc trade stations could ALSO generate these missions, giving players an incentive to find them.    We could even double the reward for every hour the order remains unfilled.  (conditional on at least one player being online of course- it would reset if the server was empty)

Basically a simple analogue of the old elite/frontier games "famine in XX star system"  which automatically increased food prices if you traded in those areas.

This could also make use of my unfinished "news system"  players could type /news   to see what current "urgent orders " have been issued by each NPC base.

Sounds pretty damn fun to me, what you think?

**Comment by jpcsupplies** [comment]
"Delivery contract bonus"  Transport tycoon style - first person gets a delivery contract for themself (and their faction?) pays extra for sells of that item for a period of time to them and theri faction?

**Comment by jpcsupplies** [comment]
See also #70 
probably need a /contracts command  to display currently held delivery subsidys
probably need a /addevent or similar command to manually create contracts in addition to those that are automatically generated by low stock (assuming we allow that to be automatic at all)
Contract/subsidys should work by two factors: 1 the time limit, 2 the buy limit, and 3 is the bonus a % on top of buy price or a fixed bonus amount., 4 should it advertise the location?

Adding contracts probably should be allowed by players who own markets too - but given they can move this might be tricky it may need to be limited only to zones that were created on a station - and will be revoked (with a penalty fee?) if the zone is moved before the contract ends.

Eg Screaming Angel has offered a delivery contract for (optional qty ( first 1000)) "missiles" delivered to base "Mutual Assured Rationality", (optional (located at GPS: x y z)) deliveries will attract a bonus (20% | 2 credits) per item, for the next (7) days.

Automatic contracts for NPC bases could probably force players to buy the goods from a particular base too (in the case of overstock vs understock?)
Eg
Blue Mining Inc desperately requires (optional qty: (50,000)) steel plates at trade facility "Hell Outpost" GPS: 1 2 3 (optional: to be transported from "Mars Outpost" GPS 3 2 1) and deliveries of this item will pay a bonus (30% | 0.05) per unit

In the case of specific forced deliveries between bases it could make use of the /load and /unload  cargo commands, and pay a bonus for the delivery only.??  Although they were more for allowing market owners/employees to add/remove stock to markets without being billed.

**Comment by jpcsupplies** [comment]
ok have a test mission up as of last commit, its just investigate 0,0,0 but seems to confirm the system is workable..   :)

**Comment by jpcsupplies** [comment]
looks like our mission system needs to cater for a mission start message a mission complete message and an option mission failure message too

**Comment by jpcsupplies** [comment]
Ok some thoughts here -
Missions should come in two desticnt flavor classes -
1: Procedural missions.   These are the automatic missions.   At the very least it should be able to: 
A) Detect if a station has low stock of a defined list of items (eg silver, platinum, iron, ice, uranium etc) create a "bonus delivery" mission for ore types that are low or exhausted.  TO npc stations only.    Because it will create a target GPS maybe allow players to manually create bonus delivery missions for their own zones.    

B) Can we scan the  "last online" data for grids.   Any grid which has not seen any of its builders in 3 months (user defined period?)  gets added to the automatic "investigate" mission rotation..   this rotation could also include NPC ships that spawn (or went adrift) far from anyone.  So we use the "investigate location X" missions randomly and pick a random abandoned grid.

C) scripted mission assigned to the "random trigger"


2: Scripted missions.  
These are configured by a settings file that triggers a mission chain.     The start triggers could be defined as specific conditions (such as the demo mission; or completion of certain prior missions, or checking for missions while in proximity of a particular npc station etc) or the chains could be started by semi random events which initiate the mission chain and offer it to someone who has indicated the wish to be advised of missions.  (ie the /mission command)
Eg the trigger could be defined as "random / if player bal > 25000"  that could trigger an ambush investigate mission (to an npc pirate or admin created grid allocated to this purpose)


These classes of missions could work independantly.   The purely procedural missions would run independant of the scripted missions..   although at a later stage the procedural system could also trigger the start of defined scripted missions..  it wont actually concern itself with managing the scripted mission system itself.
